lib/pure/options.nim

```diff
@@ -312,3 +667,126 @@ when isMainModule:
       check($noperson == "None[Person]")
+
+    test "match":
+      let x = some(100)
+      match x:
+        some y:
+          check y == 100
+        none: discard
+
+      var echoed = ""
+      proc mockEcho(input: varargs[string, `$`]) =
+        echoed = input[0]
+        for i in 1..input.high:
+          echoed = echoed & input[i]
+
+      match some(100):
+        some x: mockEcho x
+        some _ of 100: mockEcho "Is hundred"
+        none: mockEcho "No value"
+
+      check echoed == "Is hundred"
+
+      var res = match(none(int)) do:
+        some _: "Hello"
+        none: "No value"
+
+      check res == "No value"
+
+    test "compare":
+      proc leq(x, y: int): bool =
+        x <= y
+      check optCmp(10, 20, leq) == some(10)
+      check optCmp(30, 20, leq).isNone
+
+    test "logical operators":
+      let a = none[string]()
+      let b = none[string]()
+
+      let strings = ["a", "b", "c", "d", "e"]
+      var sideEffects = 0
+
+      proc genSome(): Option[string] =
+        result = some(strings[sideEffects])
+        sideEffects += 1
+
+      proc genNone(): Option[string] =
+        result = none[string]()
+        sideEffects += 1
+
+      proc reset(): void =
+        sideEffects = 0
+
+      # test ``or`` operator
+
+      check isNone((genNone() or genNone()) or genNone())
+      check(sideEffects == 3)
+      reset()
+
+      check ((genNone() or genSome()) == some("b")) == true
+      check sideEffects == 2
+      reset()
+
+      check (((genNone() or genSome()) or genSome()) == some("b")) == true
+      check sideEffects == 2
+      reset()
+
+      check ((genNone() or genSome()) or "c" == "b") == true
+      check sideEffects == 2
+      reset()
+
+      check ((genNone() or genNone()) or "c" == "c") == true
+      check sideEffects == 2
+      reset()
+
+      check (((genSome() or genNone()) or genSome()) or "c" == "a") == true
+      check(sideEffects == 1)
+      reset()
+
+      # test ``and`` operator
+
+      check (((genSome() and genSome()) and genSome()) == some("c")) == true
+      check sideEffects == 3
+      reset()
+
+      check isNone(genSome() and genNone())
+      check sideEffects == 2
+      reset()
+
+      check isNone((genSome() and genNone()) and genNone())
+      check sideEffects == 2
+      reset()
+
+      check isNone((genSome() and genNone()) and "c")
+      check sideEffects == 2
+      reset()
+
+      check (((genSome() and genSome()) and "c") == some("c")) == true
+      check sideEffects == 2
+      reset()
+
+      check isNone(((genNone() and genSome()) and genNone()) and "c")
+      check sideEffects == 1
+      reset()
+
+      # change the type during the expression and mix ``and`` with ``or``
+      check genNone() and 1 or 2 == 2 == true
+      check genSome() and 1 or 2 == 1 == true
+
+    test "conditional continuation":
+      when not compiles(some("Hello world").?find('w').echo):
+        check false
+      check (some("Hello world").?find('w') == 6)
```

Comment:
this should not compile.  ``some(6)`` and ``6`` are different types. They are not equal, and therefore should not be equal.

Reply:
All values are implicitly a `some` from the converter, which is why  this works.